File: scala/core/src/main/scala/blockchain/consensus/LocalChain.scala
package blockchain.consensus

import blockchain.BlockSourcedState
import blockchain.models.*
import blockchain.codecs.given
import blockchain.utility.given
import cats.data.EitherT
import cats.effect.{Async, Ref, Resource}
import cats.implicits.*
import cats.effect.implicits.*
import fs2.concurrent.Topic
import org.typelevel.log4cats.Logger
import org.typelevel.log4cats.slf4j.{Slf4jFactory, given}

trait LocalChain[F[_]]:
  def adopt(blockId: BlockId): F[Unit]
  def currentHead: F[BlockId]
  def genesis: F[BlockId]
  def adoptions: fs2.Stream[F, BlockId]
  def blockIdAtHeight(height: Long): F[Option[BlockId]]

object LocalChain:
  def make[F[_]: Async](
      genesisId: BlockId,
      blockHeightsBSS: BlockHeights.BSS[F],
      initialHead: BlockId
  ): Resource[F, LocalChain[F]] =
    (
      Resource.make(Topic[F, BlockId])(_.close.void),
      Ref.of(initialHead).toResource
    )
      .mapN((topic, ref) =>
        new LocalChainImpl[F](genesisId, blockHeightsBSS, topic, ref)
      )

class LocalChainImpl[F[_]: Async](
    genesisId: BlockId,
    blockHeightsBSS: BlockHeights.BSS[F],
    adoptionsTopic: Topic[F, BlockId],
    headRef: Ref[F, BlockId]
) extends LocalChain[F]:

  private given Logger[F] =
    Slf4jFactory.getLoggerFromName[F]("Blockchain.Consensus.LocalChain")
  override def adopt(blockId: BlockId): F[Unit] =
    Async[F].uncancelable(_ =>
      headRef.set(blockId) *>
        EitherT(adoptionsTopic.publish1(blockId))
          .leftMap(_ =>
            new IllegalStateException("LocalChain topic unexpectedly closed")
          )
          .rethrowT *>
        Logger[F].info(show"Adopted head block id=$blockId")
    )

  override def currentHead: F[BlockId] = headRef.get

  override def genesis: F[BlockId] = genesisId.pure[F]

  override def adoptions: fs2.Stream[F, BlockId] =
    adoptionsTopic.subscribe(Int.MaxValue)

  override def blockIdAtHeight(height: Long): F[Option[BlockId]] =
    currentHead.flatMap(blockHeightsBSS.useStateAt(_)(_.apply(height)))

File: dart/protocol/lib/src/blockchain/consensus/local_chain.dart
import 'dart:async';

import 'package:fixnum/fixnum.dart';
import 'package:giraffe_protocol/src/blockchain/consensus/block_heights.dart';
import 'package:giraffe_sdk/sdk.dart';
import 'package:logging/logging.dart';

import '../common/common.dart';
import '../genesis.dart';

abstract class LocalChain {
  Future<void> adopt(BlockId blockId);
  BlockId get head;
  BlockId get genesis;
  Stream<BlockId> get adoptions;
  Future<BlockId?> blockIdAtHeight(Int64 height);
  Future<void> close();
}

class LocalChainImpl extends LocalChain {
  @override
  final BlockId genesis;
  @override
  BlockId head;
  final Future<void> Function(BlockId) onAdopted;
  final FetchHeader fetchHeader;
  final FetchBody fetchBody;
  final BlockHeightsBSS blockHeightsBSS;

  final StreamController<BlockId> _adoptions =
      StreamController<BlockId>.broadcast();

  LocalChainImpl(
      {required this.genesis,
      required this.head,
      required this.onAdopted,
      required this.fetchHeader,
      required this.fetchBody,
      required this.blockHeightsBSS});

  @override
  Future<void> adopt(BlockId blockId) async {
    head = blockId;
    await onAdopted(blockId);
    _adoptions.add(blockId);
    final header = await fetchHeader(blockId);
    final body = await fetchBody(blockId);
    log.info(
        "Adopted blockId=${blockId.show} height=${header.height} slot=${header.slot} transactions=${body.transactionIds.map((id) => id.show).toList()}");
  }

  @override
  Stream<BlockId> get adoptions => _adoptions.stream;

  @override
  Future<BlockId?> blockIdAtHeight(Int64 height) async {
    if (height == Genesis.height) {
      return genesis;
    } else if (height > Genesis.height) {
      return blockHeightsBSS.useStateAt(head, (state) => state(height));
    } else if (height == Int64.ZERO) {
      return head;
    } else {
      final currentHeader = await fetchHeader(head);
      final targetHeight = currentHeader.height + height;
      if (targetHeight < Genesis.height) return null;
      return blockIdAtHeight(targetHeight);
    }
  }

  @override
  Future<void> close() => _adoptions.close();

  static final log = Logger("LocalChain");
}
